### PR TITLE
Implement \TextOrMath, \@secondoftwo

### PR DESCRIPTION
**Pull request by edemaine**
Partially addressing #1003.
* Parser now tells MacroExpander about mode switching.  (This seems preferable to a circular reference between Parser and MacroExpander.)
* Implement `\TextOrMath`
* Improve when we switch modes so that this actually works, in all cases except single-symbol arguments.
* Add `\@secondoftwo` to match `\@firstoftwo`.
* Add comments documenting all the conditional macros

### CODE REVIEW

File: test/katex-spec.js [hunk 1]
```diff
@@ -2672,11 +2672,61 @@ describe("A macro expander", function() {
         expect("\\@ifnextchar!{yes}{no}?!").toParseLike("no?!");
     });
 
-    it("\\@firstoftwwo should consume star but nothing else", function() {
+    it("\\@ifstar should consume star but nothing else", function() {
```
Comment (kevinbarabash) [inline test/katex-spec.js]: good catch

File: test/katex-spec.js [hunk 2]
```diff
@@ -2672,11 +2672,61 @@ describe("A macro expander", function() {
...
+    it("\\TextOrMath should work immediately after $ ends", function() {
+        expect("\\text{$\\TextOrMath{text}{math}$\\TextOrMath{text}{math}}")
+            .toParseLike("\\text{$math$text}");
+    });
+
+    it("\\TextOrMath should work in a macro", function() {
+        compareParseTree("\\mode\\text{\\mode$\\mode$\\mode}\\mode",
+            "math\\text{text$math$text}math",
+            {"\\mode": "\\TextOrMath{text}{math}"});
+    });
```
Comment (kevinbarabash) [inline test/katex-spec.js]: Great test suite!

File: test/katex-spec.js [hunk 3]
```diff
@@ -2672,11 +2672,61 @@ describe("A macro expander", function() {
...
+
+    // TODO(edemaine): This doesn't work yet.  Parses like `\text math`,
+    // which doesn't even treat all four letters as an argument.
+    //it("\\TextOrMath should work in a macro passed to \\text", function() {
+    //    compareParseTree("\\text\\mode", "\\text{text}",
+    //        {"\\mode": "\\TextOrMath{text}{math}"});
```
Comment (kevinbarabash) [inline test/katex-spec.js]: What does this currently result in?

File: src/MacroExpander.js [hunk 4]
```diff
@@ -16,13 +16,22 @@ export default class MacroExpander implements MacroContextInterface {
...
         this.lexer = new Lexer(input);
         this.macros = objectAssign({}, builtinMacros, macros);
+        this.mode = mode;
         this.stack = []; // contains tokens in REVERSE order
     }
 
+    /**
+     * Switches between "text" and "math" modes.
+     */
+    switchMode(newMode) {
```
[the remark this answers is not in the document]
Comment (edemaine) [reply src/MacroExpander.js]: Added.

File: test/katex-spec.js (hunk 3 above)
Comment (edemaine) [reply test/katex-spec.js]: It results in "m" formatted in text mode followed by "ath" formatted in math mode. So it's wrong in two ways: it thinks it's in math mode when it's actually in text mode, and the argument is just the first character instead of all four. This is a (kind of known) bug with interaction between `MacroExpander` and `Parser`.

File: src/Parser.js [hunk 5]
```diff
@@ -976,6 +989,7 @@ export default class Parser {
                     firstToken.range(lastToken, firstToken.text));
         } else {
             // Otherwise, just return a nucleus, or nothing for an optional group
+            // TODO: Not yet switching modes for single-symbol group.
```
[the remark this answers is not in the document]
Comment (edemaine) [reply src/Parser.js]: I'm not actually sure how to do it -- it's pretty challenging, because by the time `parseGroup` gets called, `nextToken` *already* has the single-symbol group, which means that it already got expanded in the wrong mode.  In the old way of doing things, we could "unget" the symbol, change modes, and reget it, but that seems challenging/impossible to do if there are side-effects/errors caused by the first execution in the wrong mode.  So I think we'd need to call `parseGroup` before actually reading the first symbol of the group, which might cause a rather big change across all of `Parser.js`.

So in summary, I'm not sure how much work it'll be.  It's surely possible, but hard.  I thought I'd commit what I have so far, and open another issue for fixing this further.  On the plus side, it took me a lot of playing around with tests before I could even get this situation to arise, and in some sense #924 is even more important than this missing case.  So maybe this PR is good to merge for now?  Up to you -- I can look at this more when I have time.

**Comment by kevinbarabash** [comment]
Code changes look good to me, but there's something causing the `Colorbox` and `DashesAndQuotes` screenshot tests to fail.

**ColorBox:**
screen shot 2017-12-19 at 8 39 02 pm
The `x` should be in typewriter.

**DashesAndQuotes:**
screen shot 2017-12-19 at 8 38 30 pm
Three dashes in a row should be an em dash.

**Comment by edemaine** [comment]
Wow, the power of tests. DashesAndQuotes was a nice easy fix pointing out incorrect mode detection for the ligatures (obvious in hindsight).  Unfortunately, the Colorbox issue is exactly the missing part of this PR, switching to text mode for a single-symbol argument (here, `x`).  So maybe I *should* try to do that before this PR gets merged; otherwise, this PR would regress behavior a bit.

**Comment by kevinbarabash** [comment]
@edemaine the following seems to be sufficient to fix the `ColorBox` screenshot test:
```
        } else {
            // Otherwise, just return a nucleus, or nothing for an optional group
            if (mode) {
                this.switchMode(mode);
            }
            const result = optional ? null : this.parseSymbol();
            if (mode) {
                this.switchMode(outerMode);
            }
            return result;
        }
```

**Comment by kevinbarabash** [comment]
There may still be other places where we need to switch modes before calling `parseSymbol`, but we can deal with them in a separate PR as necessary.

**Comment by edemaine** [comment]
@kevinbarabash Thanks for investigating!  This partially fixes the single-symbol support, basically whenever the single symbol is not a macro (not expanded by the `MacroExpander`).  I'm glad this makes it mergable (backwards-compatible); now I'll open an issue for the remaining challenge of single-symbol macro.
